### PR TITLE
API adjustments based on cupertino_http usage experience

### PR DESCRIPTION
**Pull request by brianquinlan**
- Remove support for cookies (because it is hard)
- Add the ability to set headers stored as a string-to-string mapping
- Add `close` methods to ...RequestData and ...ResponseData
- Allow some fields to be nullable

---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: pkgs/http_profile/lib/http_profile.dart [hunk 1]
```diff
@@ -82,105 +137,154 @@ final class HttpProfileRequestData {
...
-  /// profile.responseData.cookies = [
-  ///   'sessionId=abc123',
-  ///   'id=def456; Max-Age=2592000; Domain=example.com',
-  /// ];
-  /// ```
-  set cookies(List<String> value) {
-    _data['cookies'] = [...value];
-    _updated();
-  }
+  /// The body of the request.
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:286]:
```suggestion
  /// The body of the response.
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 2]
```diff
@@ -198,12 +303,46 @@ final class HttpProfileResponseData {
...
+    _checkAndUpdate();
+    if (value == null) {
+      _data.remove('headers');
+      return;
+    }
     _data['headers'] = {...value};
-    _updated();
+  }
+
+  /// The response headers where duplicate headers are represented using
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:328]:
```suggestion
  /// The response headers where duplicate headers are represented using a
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 3]
```diff
@@ -7,6 +7,54 @@ import 'dart:developer' show Service, addHttpClientProfilingData;
 import 'dart:io' show HttpClient, HttpClientResponseCompressionState;
 import 'dart:isolate' show Isolate;
 
+/// "token" as defined in RFC 2616, 2.2
+/// See https://datatracker.ietf.org/doc/html/rfc2616#section-2.2
+const _tokenChars = r"!#$%&'*+\-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ^_`"
+    'abcdefghijklmnopqrstuvwxyz|~';
+
+/// Splits comma-seperated header values.
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:15]:
```suggestion
/// Splits comma-separated header values.
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 4]
```diff
@@ -7,6 +7,54 @@ import 'dart:developer' show Service, addHttpClientProfilingData;
...
+var _headerSplitter = RegExp(r'[ \t]*,[ \t]*');
+
+/// Splits comma-seperated "Set-Cookie" header values.
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:18]:
```suggestion
/// Splits comma-separated "Set-Cookie" header values.
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 5]
```diff
@@ -7,6 +7,54 @@ import 'dart:developer' show Service, addHttpClientProfilingData;
...
+/// "Set-Cookie: lang=en; Path=/foo/,SID=x23"
+/// would both be result in `response.headers` => "lang=en; Path=/foo/,SID=x23"
+///
+/// The idea behind this regex is that ",<valid token>=" is more likely to
+/// start a new <cookie-pair> then be part of <path-value> or <extension-av>.
+///
+/// See https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1
+var _setCookieSplitter = RegExp(r'[ \t]*,[ \t]*(?=[' + _tokenChars + r']+=)');
+
+/// Splits comma-seperated header values into a [List].
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:39]:
```suggestion
/// Splits comma-separated header values into a [List].
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 6]
```diff
@@ -82,105 +137,154 @@ final class HttpProfileRequestData {
...
+      return;
+    }
+    _requestData['headers'] = {...value};
   }
 
-  set headers(Map<String, List<String>> value) {
-    _data['headers'] = {...value};
-    _updated();
+  /// The request headers where duplicate headers are represented using
+  /// comma-seperated list of values.
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:188]:
```suggestion
  /// comma-separated list of values.
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 7]
```diff
@@ -82,105 +137,154 @@ final class HttpProfileRequestData {
...
+      _requestData.remove('headers');
+      return;
+    }
+    _requestData['headers'] = {...value};
   }
 
-  set headers(Map<String, List<String>> value) {
-    _data['headers'] = {...value};
-    _updated();
+  /// The request headers where duplicate headers are represented using
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:187]:
```suggestion
  /// The request headers where duplicate headers are represented using a
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 8]
```diff
@@ -198,12 +303,46 @@ final class HttpProfileResponseData {
...
+    if (value == null) {
+      _data.remove('headers');
+      return;
+    }
     _data['headers'] = {...value};
-    _updated();
+  }
+
+  /// The response headers where duplicate headers are represented using
+  /// comma-seperated list of values.
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:329]:
```suggestion
  /// comma-separated list of values.
```

File: pkgs/http_profile/lib/http_profile.dart [hunk 9]
```diff
@@ -82,105 +137,154 @@ final class HttpProfileRequestData {
...
+  ///
+  /// [bodySink] will be closed and the fields of [HttpProfileRequestData] will
+  /// no longer be changeable.
+  ///
+  /// [endTime] is the time when the request was fully sent. It defaults to the
+  /// current time.
+  void close([DateTime? endTime]) {
+    _checkAndUpdate();
+    _isClosed = true;
+    bodySink.close();
```
Comment (derekxu16) [inline pkgs/http_profile/lib/http_profile.dart:249]: Consider changing this and the other `bodySink.close()` lines to `unawaited(bodySink.close());`

Comment (brianquinlan) [reply pkgs/http_profile/lib/http_profile.dart]: Done.

**Review by derekxu16 — APPROVED** [review]

**Pull request merged**